Added testcase for object lock

vivekanandan-k-rh opened the pull request:
Automated object lock issue

Steps:
    Create bucket with bucket lock enabled
    Enable object lock configuration
    Upload object to the bucket
    Upload another object in same bucket
    Verify new version is created for object
    Verify deletion of object versions are disabled

vivekanandan-k-rh commented:
Blocked because of https://bugzilla.redhat.com/show_bug.cgi?id=1982558

rakeshgm commented:
yaml file is missing and also tox tests failed. can you check and keep these changes ready. So once the bug gets fixed. we can just run the test.

vivekanandan-k-rh commented on rgw/v2/tests/s3_swift/test_object_lock.py:
```diff
@@ -0,0 +1,162 @@
+Usage: test_object_lock.py -c <input_yaml>
+
+<input_yaml>
+	test_object_lock_py
```
Modified it.

vivekanandan-k-rh commented:
> yaml file is missing and also tox tests failed. can you check and keep these changes ready. So once the bug gets fixed. we can just run the test.

Fixed and added file changes. Earlier commit was done just to take backup of so far completed code changes

vivekanandan-k-rh commented:
Passed log: http://pastebin.test.redhat.com/980915

rakeshgm commented on rgw/v2/tests/s3_swift/test_object_lock.py:
```diff
@@ -0,0 +1,186 @@
+    all_users_info = s3lib.create_users(config.user_count)
+    for each_user in all_users_info:
+        # authentication
+        auth = Auth(each_user, ssl=config.ssl)
+        rgw_conn = auth.do_auth()
```
Since you wanna use client function to authenticate, there is an internal function in `Auth` class for client. use that here.

vivekanandan-k-rh replied:
Done.

rakeshgm commented on rgw/v2/tests/s3_swift/test_object_lock.py:
```diff
@@ -0,0 +1,186 @@
+                )
+                log.info(f"creating bucket with name: {bucket_name_to_create}")
+                rgw_ip_and_port = get_rgw_ip_and_port()
+                s3_conn_client = boto3.client(
+                    "s3",
+                    aws_access_key_id=each_user["access_key"],
+                    aws_secret_access_key=each_user["secret_key"],
+                    endpoint_url=f"http://{rgw_ip_and_port}",
+                )
```
remove this once you use `do_auth_using_client`

vivekanandan-k-rh replied:
Done.

rakeshgm commented on rgw/v2/tests/s3_swift/test_object_lock.py:
```diff
@@ -0,0 +1,186 @@
+                    error_message = (
+                        f"Expected: {config.objects_count}, Actual: {versions_count}"
+                    )
+                    assert versions_count == config.objects_count, error_message
```
don't use asserts anywhere in the test cases. exceptions make sense. please remove all the assertion. refer other tests.

vivekanandan-k-rh replied:
Modified it.

rakeshgm commented:
I am adding DNM label for this.

psathyan commented on rgw/v2/tests/s3_swift/configs/test_object_lock.yaml:
```diff
@@ -0,0 +1,12 @@
+# upload type: non multipart
+# script: test_object_lock.py
+config:
+ user_count: 1
+ bucket_count: 1
+ objects_count: 2
+ objects_size_range:
+  min: 5
+  max: 15
+ test_ops:
+      create_bucket: true
+      create_object: true
```
It is a good YAML practice to have two spacing or a consistent spacing in the file. By this way, we can avoid any loading issues... here we could do
```
config:
  user_count: 1
  bucket_count: 1
  objects_count: 2
  objects_size_range:
    min: 5
    max: 15
  test_ops:
    create_bucket: true
    create_object: true
```

vivekanandan-k-rh replied:
Done

psathyan commented on rgw/v2/tests/s3_swift/configs/test_object_lock.yaml:
```diff
@@ -0,0 +1,12 @@
+# upload type: non multipart
+# script: test_object_lock.py
```
It's always helpful if could add meaning comment messages.

vivekanandan-k-rh replied:
Done. Updated

psathyan commented on rgw/v2/tests/s3_swift/test_object_lock.py:
```diff
@@ -0,0 +1,186 @@
+""" test_object_lock.py - Test object lock configuration for bucket
```
Nice.

psathyan commented on rgw/v2/tests/s3_swift/test_object_lock.py:
```diff
@@ -0,0 +1,186 @@
+        if not os.path.exists(TEST_DATA_PATH):
+            log.info("test data dir not exists, creating.. ")
+            os.makedirs(TEST_DATA_PATH)
+        parser = argparse.ArgumentParser(description="Object Lock Configuration")
```
Can we move to `docopt`? You have put in a very nice module description...

vivekanandan-k-rh replied:
Updated the description with usage. But not used docopt as the function exits when I use docopt.

rakeshgm replied:
docopt is a good idea. i was meaning to get most of the code to use docopt. we can definitely make docopt usage here. maybe a next PR with all the changes.
